src/main/java/com/faforever/client/map/generator/MapGeneratorServiceImpl.java

```diff
@@ -0,0 +1,143 @@
+package com.faforever.client.map.generator;
+
+import com.faforever.client.i18n.I18n;
+import com.faforever.client.io.FileUtils;
+import com.faforever.client.notification.ImmediateNotification;
+import com.faforever.client.notification.NotificationService;
+import com.faforever.client.notification.Severity;
+import com.faforever.client.notification.TransientNotification;
+import com.faforever.client.preferences.PreferencesService;
+import com.faforever.client.task.TaskService;
+import com.google.common.eventbus.EventBus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.github.nocatch.NoCatch.noCatch;
+
+@Lazy
+@Service
+public class MapGeneratorServiceImpl implements MapGeneratorService {
+
+  private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final String GENERATED_MAP_NAME = "NeroxisMapGenerator_%d_%s";
+  private static final String GENERATOR_DEFAULT_VERSION = "0.0.1-prealpha";
+  private static final String GENERATOR_EXECUTABLE_FILENAME = "MapGenerator_%s.jar";
+
+  private static final int GENERATION_TIMEOUT_SECONDS = 20;
+
+  private static final Pattern GENERATED_MAP_PATTERN = Pattern.compile("NeroxisMapGenerator_(-?\\d+)_(\\S+)");
+
+  private final PreferencesService preferencesService;
+  private final TaskService taskService;
+  private final I18n i18n;
+  private final NotificationService notificationService;
+  private final EventBus eventBus;
+
+  private Path customMapsDirectory;
+  private Random seedGenerator;
+
+  @Inject
+  public MapGeneratorServiceImpl(PreferencesService preferencesService, TaskService taskService, I18n i18n, NotificationService notificationService, EventBus eventBus) {
+    this.preferencesService = preferencesService;
+    this.taskService = taskService;
+    this.i18n = i18n;
+    this.notificationService = notificationService;
+    this.eventBus = eventBus;
+
+
+    seedGenerator = new Random();
+
+    customMapsDirectory = this.preferencesService.getPreferences().getForgedAlliance().getCustomMapsDirectory();
+    deleteGeneratedMaps();
```

Comment:
Try to delete Maps on destroying the bean avois permission problems

Reply:
You mean I should really block the clients closing procedure (`@PreDestroy`) to delete leftover maps?